### DISCUSSION TITLE
BBHFactory & Picture Recorder

### DISCUSSION THREAD

**Discussion by parkerpierpont** [rust-skia/rust-skia#562] [Q&A] [answered]
Hi!

TLDR; is this on the roadmap, alternatively, seeking direction on what I'd need to know and do to contribute/implement.

I'm looking to build a compositor and integrate it with skia-safe. I'm finally starting fo wrap my head around Rust, and it felt like a good project that could help me get more solidly into an intermediate level of proficiency with the language (as my day job is with higher level languages).

It seems like everything I would need is possible to do - but it looks like BBHFactory & BBH are both not implemented yet.

I've tried my hand at naively forking this lib to try generating these things - but I only got the point of getting the vtable errors that the author left a comment about in the skia-bindings/build-support folder. (Btw, the ergonomics of building is fantastic - and I've learned a lot by playing with the build process and bindings.)

I'm a little stumped as to how to go about getting these virtual methods to be accessible to Rust. 

I'm pretty unfamiliar with C++, is anyone have information/already implemented/similar cases in the existing code base I can use to try and implement this?

If this is on the roadmap already, I'd be happy to wait a little bit - but I'm also happy to work on a PR if there are existing examples/articles/references/google-keywords anyone could point me towards.

Alternatively, any info about how you've managed to trick bindgen into not needing to maintain a vtable for a class with inheritance/virtual functions would be greatly appreciated.

**Comment by pragmatrix** [comment, marked as answer]
> I'm looking to build a compositor and integrate it with skia-safe.

Sounds exciting!

> I'm a little stumped as to how to go about getting these virtual methods to be accessible to Rust.

In rust-skia, we ignored the VTables and created small C++ wrapper functions that invoke the virtual methods. Going through VTables wasn't considered, because seemed to rely too much on the C++ internals / ABI and there were also problems with bindgen creating broken class layouts that caused test errors when VTables were involved (that's where the comments are coming from, and why `SkBBoxHierarchy` was marked as an opaque type), but you might find other Rust projects that attempt to do this. It surely would be an interesting experiment. 

One simple example for a a virtual wrapper I found is [C_SkShader_isOpaque](https://github.com/rust-skia/rust-skia/blob/4c6ea3c1f5860d1ee7f8cf9d184406f4b7dcfd38/skia-bindings/src/bindings.cpp#L1861-L1863), which wraps `SkShader.isOpaque()` and there are a lot more wrappers in `bindings.cpp`. For every `extern "C"` function that is prefixed with `C_`, bindings are generated making the function accessible from Rust. Even without deep knowledge of C or C++, it should be possible to create these wrappers.

To dig a bit deeper, specifically the interfacing reference counted types, https://github.com/rust-skia/rust-skia/wiki/Wrapper-Types may provide some more help.

Sidenote: While we would appreciate bindings for these types, I am not sure if the right way to start a compositor is to create bindings for a specific way to organize bounding boxes. This is often a detail that can be delayed until after the compositor actually works but needs some optimization regarding the computation of intersection rectangles. Also I think that this is probably a part that you wish to see on the Rust side, because usually there may appear specific requirements for the compositor that may not be covered by Skia. The implementation of `SkRTree` seems to be rather small (under 200 lines of C++ code), which - if there is not already some crate lying around doing something similar - may be trivial to port to Rust.

**Reply by parkerpierpont** [reply]
This is perfect, @pragmatrix. 

I had initially had the idea of just rewriting the tree from BBHFactory downwards to Rust – but felt bad about creating some weird workaround solution in Rust that will need to track all of the drawn items, especially if this was to be a common solution for others (since you've done such a wonderful job of integrating some very complicated functions/classes from Skia in an elegant way).

I feel much more comfortable progressing in the matter you described above in Rust, and I'm now very excited to continue. Even if I change course again and need to create opaque types in C++, having a clear example of what that looks like when dealing with types that need to be opaque from the SkShader example is massively helpful.

> While we would appreciate bindings for these types, I am not sure if the right way to start a compositor is to create bindings for a specific way to organize bounding boxes. This is often a detail that can be delayed until after the compositor actually works but needs some optimization regarding the computation of intersection rectangles.

This is exactly where I was starting to become overwhelmed!

So much valuable info here. I love this project, and I've learned a ton about FFI with C++, as well as some interesting design patterns by diving into both skia-bindings and skia-safe over the past week or so.

Just one more **thank you**, for maintaining this project! I've learned a lot from all of the previous answers you've given in discussions/issues/comments/skia-org about integrating with gpu-backends, windowing libraries, etc. Just the build process in skia-bindings itself was a real eye-opener for me. It's been a joy to use and dig into this project.
